mjpegtools-snapshot/docs/y4mtopnm.1
.\"
.TH "y4mtopnm" "1" "28 April 2004" "MJPEG Tools Team" "MJPEG tools manual"
.SH "NAME"
y4mtopnm \- Convert a YUV4MPEG2 stream to PNM images

.SH "SYNOPSIS"
.B y4mtopnm
.RI [ options ]

.SH "DESCRIPTION"
\fBy4mtopnm\fP converts a YUV4MPEG2 stream into a sequence
of raw PPM, PGM, or PAM images.  (Technically, "pnm" comprises the PPM, PGM and
PBM formats.  PAM is a fourth format, which can effectively contain any of
the other three, and more.)

Output is to stdout (but feel free to have the shell redirect to a file).

Input is read from stdin, like all other YUV4MPEG2 filters and tools.
YUV4MPEG2 streams contain frames using the Y'CbCr colorspace
(ITU-R BT.601).  For 4:4:4 (color) streams, \fBy4mtopnm\fP will convert
each pixel to the usual R'G'B' colorspace used for computer graphics,
and produce PPM images.  "MONO" (luma-only) streams will be converted
to full-range [0,255] grayscale and output as PGM images.

YUV4MPEG2 streams may (often!) have subsampled chroma planes, but
\fBy4mtopnm\fP will not process them (except in the special 'flatten' mode;
see below).  Subsampled streams will need to
be converted to 4:4:4 using a tool such as \fBy4mscaler\fP.

If \fBy4mtopnm\fP is given a "444ALPHA" stream (4:4:4 video with
an 8-bit alpha channel), the results depend on whether or not the "-P" option
is specified.  With "-P", it will produce PAM images with a TUPLTYPE of
RGB_ALPHA, containing both color and alpha channel data.
Without "-P", it will produce pairs of PPM and PGM images:
a PPM for the color pixels followed by a PGM for the alpha channel data.
(Such a sequence of interspersed PPM and PGM images cannot be later
re-processed by \fBpnmtoy4m\fP.  This feature is mostly for quick and
easy stream debugging.)

If multiple output images are generated, they are simply output one after
another.  If you want to turn such a "multi-image" stream/file into
individual files, use \fBpnmsplit\fP.  (Some PNM filters can process
multi-image files/streams; however, many written before June 2000 will
only process the first image.)

\fBy4mtopnm\fP and \fBpnmtoy4m\fP are inverses of each other; you can
typically pipe the output of one into the other, and vice-versa (except
for the "444ALPHA" case noted above and the "mixed-mode" case noted below).
Note that the colorspace operations
can be lossy in both directions (due to range-compression and/or
quantization).  And, when converting to PNM, information on
interlacing and sample aspect ratio of a stream
is lost (but can be reconstructed by supplying command-line arguments to
\fBpnmtoy4m\fP).


.SH "OPTIONS"
\fBy4mtopnm\fP accepts the following options:

.TP 5
.BI \-P
Produce PAM format output, instead of PPM and/or PGM.  PAM can encode a
superset of PNM, but is not yet understood by as many graphics programs.
(Indeed, surprisingly few of the NetPBM tools seem to understand it.)
.TP 5
.BI \-D
Produce a separate output image for each field of each frame.
(Otherwise, a single image per frame, containing two interleaved fields,
is created.)

If the input stream is interlaced, the field-images are output in temporal
order, i.e. the first image of a bottom-field-first stream will be the
bottom-field.
If the stream is progressive or has "mixed-mode" interlacing, then the
temporal order is ambiguous, and the top-field will always be output first.

Note that it will \fInot\fP be possible to later use \fBpnmtoy4m\fP to
reconstruct a "mixed-mode" stream from a sequence of PNM/PAM images.
.TP 5
.BI \-f
Produce a special 'flattened' output, most useful for analyzing and
debugging streams.  All the planes of a frame/field are tiled together into
an enlarged PGM (or grayscale PAM) output image as follows:
.nf
               +-----+                    +-----+
       4:4:4   |  Y  |    4:4:4 w/alpha   |  Y  |
               +-----+                    +-----+
               |  Cb |                    |  Cb |
               +-----+                    +-----+
               |  Cr |                    |  Cr |
               +-----+                    +-----+
                                          |  A  |
               +-----+                    +-----+
       4:2:2,  |  Y  |
       4:2:0   +--+--+                    +-----+--+--+
               |Cb|Cr|            4:1:1   |  Y  |Cb|Cr|
               +--+--+                    +-----+--+--+
                            +-----+
                     MONO   |  Y  |
                            +-----+
.fi
In this mode, any YUV4MPEG2 chroma format is accepted, but no upsampling or
colorspace conversion is performed.
This is essentially just a reformatting/permutation of the original data from
the input stream into a PGM image container.
All the other command-line flags continue to have the same effect on output.
.TP 5
.BI \-v " [0,1,2]"
Set verbosity level.
 0 = warnings and errors only.
 1 = add informative messages, too.
 2 = add chatty debugging message, too.

.SH "EXAMPLES"
.hw y4mtopnm pnmsplit lav2yuv
To turn the first 15 frames of an (MJPEG or DV) AVI file into individual
PPM files:

.RS 5
lav2yuv -f 15 your-video.avi | y4mscaler -O chromass=444 | y4mtopnm | pnmsplit - "your-video-%d.ppm"
.RE

.SH "AUTHOR"
This manual page was written by Matt Marjanovic.
.br
If you have questions, remarks, problems or you just want to contact
the developers, the main mailing list for the MJPEG\-tools is:
  \fImjpeg\-users@lists.sourceforge.net\fP

.TP
For more info, see our website at
.I http://mjpeg.sourceforge.net/

.SH "SEE ALSO"
.BR pam (5),
.BR pgm (5),
.BR pnm (5),
.BR ppm (5),
.BR pnmtoy4m (1),
.BR mjpegtools (1),
.BR mpeg2enc (1),
.BR lav2yuv (1),
.BR pnmsplit (1),
.BR y4mscaler (1)


NEWS
Release History

   dvda-author-10.06
          June    14, 2010  Major fixes for bugs and regressions. Portability enhancements.

   dvda-author-09.09
	  Nov     30, 2009  User-defined and automatic submenus, continuous and hierarchical.
	  Oct     01, 2009  New build system and pruned source package. Release candidate 2.
	  Sep     01, 2009  Experimental version with top-menu and still picture authoring
                            and new audio options. Added to SF git repository.

   dvda-author-09.05
          May     27, 2009  Stabilized release added SoX interface.

   dvda-author-08.12
          Dec     25, 2008  First release with multichannel support and multiformat support.

   dvda-author-08.09-dev
          Sept    20, 2008  First release.

   dvda-author-08.08-dev
          August  20, 2008  Source code of 08-08-dev made portable to *nix platforms.
          August  10, 2008  First release.

   dvda-author-08.07
          Sept    01, 2008  Build 11 released.
          August  23, 2008  Build 8 released with cumulative fixes for portability
          August  02, 2008  Made available on http://dvd-audio.sourceforge.net
          July    30, 2008  Corrected bugs, see ChangeLog.

   dvda-author-08.06
          July    04, 2008  Made available on http://dvd-audio.sourceforge.net
          June    29, 2008  Finalized tests.
          June        2008  New developments by Fabrice Nicol.

   dvda-author-20050703
          July    03, 2005  Latest Dave Chapman source code.



mjpegtools-2.1.0/lavtools/y4mtopnm.c
/*
 * y4mtopnm.c:  Convert a YUV4MPEG2 stream into one or more PPM/PGM/PAM images.
 *
 *              Converts ITU-Rec.601 Y'CbCr to R'G'B' colorspace
 *               (or Rec.601 Y' to [0,255] grayscale, for PGM).
 *
 *
 *  Copyright (C) 2005 Matthew J. Marjanovic <maddog@mir.com>
 *
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

#include "config.h"

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <assert.h>

#include <yuv4mpeg.h>
#include <mpegconsts.h>
#include "colorspace.h"


/* command-line parameters */
typedef struct _cl_info {
  int make_pam;
  int make_flat;
  int deinterleave;
  int verbosity;
  FILE *outfp;
} cl_info_t;



static
void usage(const char *progname)
{
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:  %s [options]\n", progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "Reads YUV4MPEG2 stream from stdin and produces RAW PPM, PGM\n");
  fprintf(stderr, " or PAM image(s) on stdout.  Converts digital video Y'CbCr colorspace\n");
  fprintf(stderr, " to computer graphics R'G'B'.\n");
  fprintf(stderr, "\n");
  fprintf(stderr, " options:  (defaults specified in [])\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "  -P    produce PAM output instead of PPM/PGM\n");
  fprintf(stderr, "  -D    de-interleave fields into two PNM images\n");
  fprintf(stderr, "  -f    'flatten' multiplane frames/fields in raw composite grayscale images\n");
  fprintf(stderr, "          (for stream debugging, see manpage for details)\n");
  fprintf(stderr, "  -v n  verbosity (0,1,2) [1]\n");
  fprintf(stderr, "  -h    print this help message\n");
}



static
void parse_args(cl_info_t *cl, int argc, char **argv)
{
  int c;

  cl->make_pam = 0;
  cl->make_flat = 0;
  cl->deinterleave = 0;
  cl->verbosity = 1;
  cl->outfp = stdout; /* default to stdout */

  while ((c = getopt(argc, argv, "PDfv:h")) != -1) {
    switch (c) {
    case 'P':
      cl->make_pam = 1;
      break;
    case 'D':
      cl->deinterleave = 1;
      break;
    case 'f':
      cl->make_flat = 1;
      break;
    case 'v':
      cl->verbosity = atoi(optarg);
      if ((cl->verbosity < 0) || (cl->verbosity > 2))
	mjpeg_error("Verbosity must be 0, 1, or 2:  '%s'", optarg);
      break;
    case 'h':
      usage(argv[0]);
      exit(0);
      break;
    case '?':
    default:
      goto ERROR_EXIT;
      break;
    }
  }
  if (optind != argc) 
    goto ERROR_EXIT;

  mjpeg_default_handler_verbosity(cl->verbosity);
  /* DONE! */
  return;

 ERROR_EXIT:
  mjpeg_error("For usage hints, use option '-h'.  Please take a hint.");
  exit(1);

}


/* Write a single PAM image. */
static
void write_pam(FILE *fp,
               int chroma,
               uint8_t *buffers[],
               uint8_t *rowbuffer,
               int width, int height) 
{
  mjpeg_debug("write raw PAM image from one buffer, %dx%d", width, height);
  fprintf(fp, "P7\n");
  fprintf(fp, "WIDTH %d\n", width);
  fprintf(fp, "HEIGHT %d\n", height);
  fprintf(fp, "MAXVAL 255\n");
  switch (chroma) {
  case Y4M_CHROMA_444:
    fprintf(fp, "DEPTH 3\n");
    fprintf(fp, "TUPLTYPE RGB\n");
    fprintf(fp, "ENDHDR\n");
    {
      int x, y;
      uint8_t *pixels;
      uint8_t *R = buffers[0];
      uint8_t *G = buffers[1];
      uint8_t *B = buffers[2];
      assert(rowbuffer != NULL);
      for (y = 0; y < height; y++) {
        pixels = rowbuffer;
        for (x = 0; x < width; x++) {
          *(pixels++) = *(R++);
          *(pixels++) = *(G++);
          *(pixels++) = *(B++);
        }
        fwrite(rowbuffer, sizeof(rowbuffer[0]), width * 3, fp);
      }
    }
    break;
  case Y4M_CHROMA_444ALPHA:
    fprintf(fp, "DEPTH 4\n");
    fprintf(fp, "TUPLTYPE RGB_ALPHA\n");
    fprintf(fp, "ENDHDR\n");
    {
      int x, y;
      uint8_t *pixels;
      uint8_t *R = buffers[0];
      uint8_t *G = buffers[1];
      uint8_t *B = buffers[2];
      uint8_t *A = buffers[3];
      assert(rowbuffer != NULL);
      for (y = 0; y < height; y++) {
        pixels = rowbuffer;
        for (x = 0; x < width; x++) {
          *(pixels++) = *(R++);
          *(pixels++) = *(G++);
          *(pixels++) = *(B++);
          *(pixels++) = *(A++);
        }
        fwrite(rowbuffer, sizeof(rowbuffer[0]), width * 4, fp);
      }
    }
    break;
  case Y4M_CHROMA_MONO:
    fprintf(fp, "DEPTH 1\n");
    fprintf(fp, "TUPLTYPE GRAYSCALE\n");
    fprintf(fp, "ENDHDR\n");
    fwrite(buffers[0], sizeof(buffers[0][0]), width * height, fp);
    break;
  }
  
}


/* Write a single PPM image. */
static
void write_ppm(FILE *fp,
               uint8_t *buffers[],
               uint8_t *rowbuffer,
               int width, int height) 
{
  int x, y;
  uint8_t *pixels;
  uint8_t *R = buffers[0];
  uint8_t *G = buffers[1];
  uint8_t *B = buffers[2];

  mjpeg_debug("write raw PPM image from one buffer, %dx%d", width, height);
  fprintf(fp, "P6\n%d %d 255\n", width, height);
  assert(rowbuffer != NULL);
  for (y = 0; y < height; y++) {
    pixels = rowbuffer;
    for (x = 0; x < width; x++) {
      *(pixels++) = *(R++);
      *(pixels++) = *(G++);
      *(pixels++) = *(B++);
    }
    fwrite(rowbuffer, sizeof(rowbuffer[0]), width * 3, fp);
  }
}


/* Write a single PGM image. */
static
void write_pgm(FILE *fp, uint8_t *buffer, int width, int height) 
{
  mjpeg_debug("write raw PGM image from one buffer, %dx%d", width, height);
  fprintf(fp, "P5\n%d %d 255\n", width, height);
  fwrite(buffer, sizeof(buffer[0]), width * height, fp);
}



/* Write a frame or field out as "an image".
   A frame/field with an alpha channel may become two images... */
static
void write_image(FILE *fp,
                 int make_pam,
                 int chroma,
                 uint8_t *buffers[],
                 uint8_t *rowbuffer,
                 int width, int height) 
{
  if (make_pam) {
    write_pam(fp, chroma, buffers, rowbuffer, width, height);
  } else {
    switch (chroma) {
    case Y4M_CHROMA_444:
      write_ppm(fp, buffers, rowbuffer, width, height);
      break;
    case Y4M_CHROMA_444ALPHA:
      write_ppm(fp, buffers, rowbuffer, width, height);
      write_pgm(fp, buffers[3], width, height);
      break;
    case Y4M_CHROMA_MONO:
      write_pgm(fp, buffers[0], width, height);
      break;
    default:
      assert(0);  break;
    }
  }
}

/************************************************************************/
/************** Structure/code for 'flattening' images ******************/
/************************************************************************/

typedef struct {
  int chroma;
  int planes;
  int image_w;
  int image_h;
  int flat_w;
  int flat_h;
  uint8_t *flat_buffer;
} flattener_t;

static
void init_flattener(flattener_t *fl,
                    int chroma, int planes, int width, int height)
{
  fl->chroma = chroma;
  fl->planes = planes;
  fl->image_w = width;
  fl->image_h = height;
  switch (chroma) {
  case Y4M_CHROMA_444:        
  case Y4M_CHROMA_444ALPHA:   
  case Y4M_CHROMA_MONO:       
  case Y4M_CHROMA_420JPEG:
  case Y4M_CHROMA_420MPEG2:
  case Y4M_CHROMA_420PALDV:   
  case Y4M_CHROMA_422:        fl->flat_w = width;           break;
  case Y4M_CHROMA_411:        fl->flat_w = 3 * width / 2;   break;
  default:
    assert(0); break;
  }
  switch (chroma) {
  case Y4M_CHROMA_444:        fl->flat_h = 3 * height;      break;
  case Y4M_CHROMA_444ALPHA:   fl->flat_h = 4 * height;      break; 
  case Y4M_CHROMA_MONO:       fl->flat_h = height;          break;
  case Y4M_CHROMA_420JPEG:
  case Y4M_CHROMA_420MPEG2:
  case Y4M_CHROMA_420PALDV:   fl->flat_h = 3 * height / 2;  break;
  case Y4M_CHROMA_422:        fl->flat_h = 2 * height;      break;
  case Y4M_CHROMA_411:        fl->flat_h = height;          break;
  default:
    assert(0); break;
  }
  fl->flat_buffer = (uint8_t *) malloc(fl->flat_w * fl->flat_h);
  assert(fl->flat_buffer != NULL);
}

static
void fini_flattener(flattener_t *fl)
{
  if (fl->flat_buffer != NULL) {
    free(fl->flat_buffer);
    fl->flat_buffer = NULL;
  }
}


static
void flatten(flattener_t *fl, uint8_t *img_buffers[])
{
  switch (fl->chroma) {
    /* Non-subsampled:  Y, Cb, Cr, (A) vertically appended */
  case Y4M_CHROMA_444ALPHA:
  case Y4M_CHROMA_444:
  case Y4M_CHROMA_MONO:
    {
      int planesize = fl->image_w * fl->image_h;
      uint8_t *fbuf = fl->flat_buffer;
      int p;
      for (p = 0; p < fl->planes; p++) {
        memcpy(fbuf, img_buffers[p], planesize);
        fbuf += planesize;
      }
    }
    break;
    /* 1/2 horizontal:  Y followed by (Cb,Cr) */
  case Y4M_CHROMA_422:
    {
      uint8_t *fbuf = fl->flat_buffer;
      uint8_t *ib1 = img_buffers[1];
      uint8_t *ib2 = img_buffers[2];
      int planesize = fl->image_w * fl->image_h;
      int halfwidth = fl->image_w / 2;
      int y;
      memcpy(fbuf, img_buffers[0], planesize);
      fbuf += planesize;
      for (y = 0; y < fl->image_h; y++) {
        memcpy(fbuf, ib1, halfwidth);
        fbuf += halfwidth;
        ib1 += halfwidth;
        memcpy(fbuf, ib2, halfwidth);
        fbuf += halfwidth;
        ib2 += halfwidth;
      }
    }
    break;
    /* 1/2 horizontal, 1/2 vertical:  Y followed by (Cb,Cr) */
  case Y4M_CHROMA_420JPEG:
  case Y4M_CHROMA_420MPEG2:
  case Y4M_CHROMA_420PALDV:
    {
      uint8_t *fbuf = fl->flat_buffer;
      uint8_t *ib1 = img_buffers[1];
      uint8_t *ib2 = img_buffers[2];
      int planesize = fl->image_w * fl->image_h;
      int halfwidth = fl->image_w / 2;
      int y;
      memcpy(fbuf, img_buffers[0], planesize);
      fbuf += planesize;
      for (y = 0; y < fl->image_h / 2; y++) {
        memcpy(fbuf, ib1, halfwidth);
        fbuf += halfwidth;
        ib1 += halfwidth;
        memcpy(fbuf, ib2, halfwidth);
        fbuf += halfwidth;
        ib2 += halfwidth;
      }
    }
    break;
    /* 1/4 horizontal:  Y, Cb, Cr horizontally appended */
  case Y4M_CHROMA_411:
    {
      uint8_t *fbuf = fl->flat_buffer;
      uint8_t *ib0 = img_buffers[0];
      uint8_t *ib1 = img_buffers[1];
      uint8_t *ib2 = img_buffers[2];
      int qtrwidth = fl->image_w / 4;
      int y;
      for (y = 0; y < fl->image_h; y++) {
        memcpy(fbuf, ib0, fl->image_w);
        fbuf += fl->image_w;
        ib0 += fl->image_w;
        memcpy(fbuf, ib1, qtrwidth);
        fbuf += qtrwidth;
        ib1 += qtrwidth;
        memcpy(fbuf, ib2, qtrwidth);
        fbuf += qtrwidth;
        ib2 += qtrwidth;
      }
    }
    break;
  default:
    assert(0); break;
  }
}


static
void write_flattened(flattener_t *fl, FILE *fp, int make_pam)
{
  write_image(fp, make_pam, Y4M_CHROMA_MONO,
              &(fl->flat_buffer), NULL /* rowbuffer */,
              fl->flat_w, fl->flat_h);
}



static
void convert_chroma(int chroma, uint8_t *buffers[], int width, int height)
{
  switch (chroma) {
  case Y4M_CHROMA_444:
    convert_YCbCr_to_RGB(buffers, width * height);
    break;
  case Y4M_CHROMA_444ALPHA:
    convert_YCbCr_to_RGB(buffers, width * height);
    convert_Y219_to_Y255(buffers[3], width * height);
    break;
  case Y4M_CHROMA_MONO:
    convert_Y219_to_Y255(buffers[0], width * height);
    break;
  default:
    assert(0);  break;
  }
}



int main(int argc, char **argv)
{
  int in_fd = 0;
  cl_info_t cl;
  y4m_stream_info_t streaminfo;
  y4m_frame_info_t frameinfo;
  uint8_t *buffers[Y4M_MAX_NUM_PLANES];  /* R'G'B' or Y'CbCr */
  uint8_t *buffers2[Y4M_MAX_NUM_PLANES]; /* R'G'B' or Y'CbCr */
  int err, i;
  int width, height;
  int interlace, chroma, planes;
  uint8_t *rowbuffer = NULL;
  int frame_count;
  flattener_t fl;

  y4m_accept_extensions(1); /* allow non-4:2:0 chroma */
  y4m_init_stream_info(&streaminfo);
  y4m_init_frame_info(&frameinfo);

  parse_args(&cl, argc, argv);

  if ((err = y4m_read_stream_header(in_fd, &streaminfo)) != Y4M_OK) {
    mjpeg_error("Couldn't read YUV4MPEG2 header: %s!", y4m_strerr(err));
    exit(1);
  }
  mjpeg_info("Input stream parameters:");
  y4m_log_stream_info(mjpeg_loglev_t("info"), "<<<", &streaminfo);

  width = y4m_si_get_width(&streaminfo);
  height = y4m_si_get_height(&streaminfo);
  interlace = y4m_si_get_interlace(&streaminfo);
  chroma = y4m_si_get_chroma(&streaminfo);
  planes = y4m_si_get_plane_count(&streaminfo);

  switch (chroma) {
  case Y4M_CHROMA_444:
  case Y4M_CHROMA_444ALPHA:
  case Y4M_CHROMA_MONO:
    break;
  default:
    if (!cl.make_flat) {
      mjpeg_error("Cannot handle input stream's chroma mode!");
      mjpeg_error_exit1("Input must be non-subsampled (e.g. 4:4:4).");
    }
  }
  
  /*** Tell the user what is going to happen ***/
  mjpeg_info("Output image parameters:");
  mjpeg_info("   format:  %s",
             (cl.make_pam) ? "PAM" :
             (cl.make_flat) ? "PGM" :
             (chroma == Y4M_CHROMA_444) ? "PPM" :
             (chroma == Y4M_CHROMA_444ALPHA) ? "PPM+PGM" :
             (chroma == Y4M_CHROMA_MONO) ? "PGM" : "???");
  mjpeg_info("  framing:  %s",
             (cl.deinterleave) ? "two images per frame (deinterleaved)" :
             "one image per frame (interleaved)");
  if (cl.deinterleave) {
    mjpeg_info("    order:  %s",
               (interlace == Y4M_ILACE_BOTTOM_FIRST) ? "bottom field first" :
               "top field first");
  }
  if (cl.make_flat) 
    mjpeg_info(" 'flattened' --> all planes in one image");

  /*** Allocate buffers ***/
  mjpeg_debug("allocating buffers...");
  if (cl.make_flat) {
    if (!cl.deinterleave) 
      init_flattener(&fl, chroma, planes, width, height);
    else 
      init_flattener(&fl, chroma, planes, width, height / 2);
    rowbuffer = NULL;
  } else {
    rowbuffer = malloc(width * planes * sizeof(rowbuffer[0]));
  }
  mjpeg_debug("  rowbuffer %p", rowbuffer);
  for (i = 0; i < planes; i++) {
    switch (interlace) {
    case Y4M_ILACE_NONE:
      buffers[i] = malloc(width * height * sizeof(buffers[i][0]));
      buffers2[i] = NULL;
      break;
    case Y4M_ILACE_TOP_FIRST:
    case Y4M_ILACE_BOTTOM_FIRST:
    case Y4M_ILACE_MIXED:
      /* 'buffers' may hold whole frame or one field... */
      buffers[i] = malloc(width * height * sizeof(buffers[i][0]));
      /* ... but 'buffers2' will never hold more than one field. */
      buffers2[i] = malloc(width * height / 2 * sizeof(buffers[i][0]));
      break;
    default:
      assert(0);
      break;
    }
    mjpeg_debug("  buffers[%d] %p   buffers2[%d] %p", 
                i, buffers[i], i, buffers2[i]);
  }

  /*** Process frames ***/
  frame_count = 0;
  while (1) {
    err = y4m_read_frame_header(in_fd, &streaminfo, &frameinfo);
    if (err != Y4M_OK) break;

    frame_count++;

    if (!cl.deinterleave) {
      mjpeg_debug("reading whole frame...");
      err = y4m_read_frame_data(in_fd, &streaminfo, &frameinfo, buffers);
      if (err != Y4M_OK) break;
      if (cl.make_flat) {
        flatten(&fl, buffers);
        write_flattened(&fl, cl.outfp, cl.make_pam);
      } else {
        convert_chroma(chroma, buffers, width, height);
        write_image(cl.outfp, cl.make_pam,
                    chroma, buffers, rowbuffer, width, height);
      }
    } else {
      mjpeg_debug("reading separate fields...");
      err = y4m_read_fields_data(in_fd, &streaminfo, &frameinfo, 
                                 buffers, buffers2);
      if (err != Y4M_OK) break;
      if (cl.make_flat) {
        switch (interlace) {
        case Y4M_ILACE_NONE:       /* ambiguous temporal order */
        case Y4M_ILACE_MIXED:      /* ambiguous temporal order */
        case Y4M_ILACE_TOP_FIRST:
          /* write top field first */
          flatten(&fl, buffers);
          write_flattened(&fl, cl.outfp, cl.make_pam);
          flatten(&fl, buffers2);
          write_flattened(&fl, cl.outfp, cl.make_pam);
          break;
        case Y4M_ILACE_BOTTOM_FIRST:
          /* write bottom field first */
          flatten(&fl, buffers2);
          write_flattened(&fl, cl.outfp, cl.make_pam);
          flatten(&fl, buffers);
          write_flattened(&fl, cl.outfp, cl.make_pam);
          break;
        default:
          mjpeg_error_exit1("Unknown input interlacing mode (%d).\n",
                            interlace);
          break;
        }
      } else {
        convert_chroma(chroma, buffers, width, height / 2);
        convert_chroma(chroma, buffers2, width, height / 2);
        switch (interlace) {
        case Y4M_ILACE_NONE:       /* ambiguous temporal order */
        case Y4M_ILACE_MIXED:      /* ambiguous temporal order */
        case Y4M_ILACE_TOP_FIRST:
          /* write top field first */
          write_image(cl.outfp, cl.make_pam,
                      chroma, buffers, rowbuffer, width, height / 2);
          write_image(cl.outfp, cl.make_pam,
                      chroma, buffers2, rowbuffer, width, height / 2);
          break;
        case Y4M_ILACE_BOTTOM_FIRST:
          /* write bottom field first */
          write_image(cl.outfp, cl.make_pam,
                      chroma, buffers2, rowbuffer, width, height / 2);
          write_image(cl.outfp, cl.make_pam,
                      chroma, buffers, rowbuffer, width, height / 2);
          break;
        default:
          mjpeg_error_exit1("Unknown input interlacing mode (%d).\n",
                            interlace);
          break;
        }
      }
    }
  }     
  if ((err != Y4M_OK) && (err != Y4M_ERR_EOF))
    mjpeg_error("Couldn't read frame:  %s", y4m_strerr(err));

  mjpeg_info("Processed %d frames.", frame_count);

  /*** Clean-up after ourselves ***/
  mjpeg_debug("freeing buffers; cleaning up");
  if (rowbuffer != NULL) 
    free(rowbuffer);
  for (i = 0; i < planes; i++) {
    free(buffers[i]);
    if (buffers2[i] != NULL)
      free(buffers2[i]);
  }
  y4m_fini_stream_info(&streaminfo);
  y4m_fini_frame_info(&frameinfo);
  if (cl.make_flat)
    fini_flattener(&fl);
  
  mjpeg_debug("Done.");
  return 0;
}


